src/main/java/ch/njol/skript/expressions/ExprVehicle.java

```diff
@@ -18,186 +18,151 @@
+				return vehicleExitEvent.getVehicle();
 			}
-			if (
-				(HAS_OLD_MOUNT_EVENTS || HAS_NEW_MOUNT_EVENTS)
-				&& getTime() >= 0 && !Delay.isDelayed(e)
-				&& e instanceof EntityEvent && entity.equals(((EntityEvent) e).getEntity())
-			) {
-				if (HAS_NEW_MOUNT_EVENTS) {
-					if (e instanceof EntityMountEvent)
-						return ((EntityMountEvent) e).getMount();
-					if (e instanceof EntityDismountEvent)
-						return ((EntityDismountEvent) e).getDismounted();
-				} else { // legacy mount event support
-					try {
-						assert OLD_MOUNT_EVENT_CLASS != null;
-						if (OLD_MOUNT_EVENT_CLASS.isInstance(e)) {
-							assert OLD_GETMOUNT_HANDLE != null;
-							return (Entity) OLD_GETMOUNT_HANDLE.invoke(e);
-						}
-						assert OLD_DISMOUNT_EVENT_CLASS != null;
-						if (OLD_DISMOUNT_EVENT_CLASS.isInstance(e)) {
-							assert OLD_GETDISMOUNTED_HANDLE != null;
-							return (Entity) OLD_GETDISMOUNTED_HANDLE.invoke(e);
-						}
-					} catch (Throwable ex) {
-						Skript.exception(ex, "An error occurred while trying to invoke legacy mount event support.");
-					}
-				}
+			if (getTime() >= 0 && event instanceof EntityMountEvent entityMountEvent && entity.equals(entityMountEvent.getEntity())) {
+				return entityMountEvent.getMount();
+			}
+			if (getTime() <= 0 && event instanceof EntityDismountEvent entityDismountEvent && entity.equals(entityDismountEvent.getEntity())) {
+				return entityDismountEvent.getDismounted();
 			}
 			return entity.getVehicle();
 		});
 	}
-	
+
 	@Override
 	@Nullable
-	public Entity convert(final Entity e) {
-		assert false;
-		return e.getVehicle();
-	}
-	
-	@Override
-	public Class<? extends Entity> getReturnType() {
-		return Entity.class;
+	public Entity convert(Entity entity) {
+		return entity.getVehicle();
 	}
-	
-	@Override
-	protected String getPropertyName() {
-		return "vehicle";
-	}
-	
+
 	@Override
 	@Nullable
-	public Class<?>[] acceptChange(final ChangeMode mode) {
+	public Class<?>[] acceptChange(ChangeMode mode) {
 		if (mode == ChangeMode.SET) {
+			if (isDefault() && getParser().isCurrentEvent(VehicleExitEvent.class, EntityDismountEvent.class)) {
+				Skript.error("Setting the vehicle during a dismount/exit vehicle event will create an infinite mounting loop.");
+				return null;
+			}
 			return new Class[] {Entity.class, EntityData.class};
 		}
 		return super.acceptChange(mode);
 	}
-	
+
 	@Override
-	public void change(final Event e, final @Nullable Object[] delta, final ChangeMode mode) {
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
 		if (mode == ChangeMode.SET) {
-			assert delta != null;
-			final Entity[] ps = getExpr().getArray(e);
-			if (ps.length == 0)
+			// The player can desync if setting an entity as it's currently mounting it.
+			// Remember that there can be other entity types aside from players, so only cancel this for players.
+			Predicate<Entity> predicate = Player.class::isInstance;
```

Comment:
what's the point of doing this instead of `x instanceof Player`

Reply:
The usage of the same predicate 6 times

Reply:
yes but how is making it a predicate object doing anything but obscuring what it's intended to do
I would rather see the instanceof written out, but i'd settle for at least a descriptive name like `isInstanceOfPlayer`